pkg/ccl/crosscluster/streamclient/partitioned_stream_client.go

```diff
@@ -118,6 +118,21 @@ func (p *partitionedStreamClient) Heartbeat(
 
 	p.mu.Lock()
 	defer p.mu.Unlock()
+
+	// The heartbeat function is called in a loop with the same connection. If
+	// that connection closes, attempt to re-establish it to prevent the hearbeat
+	// loop from failing repeatedly.
+	if p.mu.srcConn.IsClosed() {
+		conn, err := pgx.ConnectConfig(ctx, p.pgxConfig)
+		if err != nil {
+			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error reconnecting to replication stream %d", streamID)
+		}
+		if err := p.mu.srcConn.Ping(ctx); err != nil {
+			return streampb.StreamReplicationStatus{}, errors.Wrapf(err, "error pinging on reconnecting replication stream %d", streamID)
```

Comment:
Do we need to close conn on this path? Alternatively, if you immediately save the conn to p.mu.src conn, perhaps we can depend on (*partitionedStreamClient).Close

Reply:
i think we should depend on close. I don't think the Heartbeat function should be closing the conn.